src/main/java/stelnet/commodity/view/TableViewFactory.java

```diff
@@ -23,31 +23,39 @@ public AbstractRenderable createContainer(String commodityId, CommodityTab activ
         float height = size.getHeight() - 35;
         float tabsHeight = 15f;
         float tableHeight = height - tabsHeight;
-        HorizontalViewContainer tabs = createTabs(activeTab);
-        TableContent tableContent = createTableContent(commodityId, activeTab);
-        Table table = new Table(commodityId, width, tableHeight, tableContent);
-        return new VerticalViewContainer(tabs, table);
+        MarketRepository marketRepository = new MarketRepository(commodityId);
+        TabViewContainer tabViewContainer = new TabViewContainer();
+        addBuyTab(tabViewContainer, commodityId, activeTab, width, tableHeight, marketRepository.getBuyMarkets());
+        addSellTab(tabViewContainer, commodityId, activeTab, width, tableHeight, marketRepository.getSellMarkets());
+        addProfitTab(tabViewContainer, commodityId, activeTab, width, tableHeight, marketRepository);
+        return tabViewContainer;
```

Comment:
I think looking back to how tabs were created before vs how they are done now then this is very good outcome. Since you are looking for perfection then I will point out a design improvement which could be done here (even though it is already better than it was before).

Whenever you want to add a new tab, this class will have to change. This goes against "closed for modification open for extension" principle.

However, as a software engineer you also have to think about when battles are worth fighting and when they are not. Atm I wouldn't go further myself.

Reply:
I also think at the current mod size (and the fact it is a plugin for a game, not a game itself) this is the right place and level of  coupling that gives most benefits vs effort.

Hypothetically speaking, how would you go on from there? I'd change each `addXTab` as own class, have those implement same interface (returns Renderable), inject them as a list and compose tabs from there. I will need to define a list of tabs at one point and the only way to get away from modifying one of the consumer classes is to add implementation details in the form of a config - define tab class names that are passed to this factory c-tor. Since reflection in mods is frown upon (Alex tries to prevent usage of it, at least for base game files) I don't think this is viable approach.